File: raw/ui/teach/transport.py
"""Playback transport for a lesson: render, play, loop, and report position.

Position is not read back from the audio device; like RAW's controller it is
derived from a wall clock started when playback began. That is accurate to a
few milliseconds, which is plenty for moving a cursor and lighting keys.

A looped practice section plays as two stages: the count-in plus the first
pass once, then the section alone with the backend looping it, so the student
hears the count-in exactly once.
"""

from __future__ import annotations

import time
from dataclasses import replace

from PyQt6.QtCore import QObject, QTimer, pyqtSignal

from ...audio.buffer import AudioBuffer
from ...audio.playback import get_backend
from ...teach.player import NoteCache, PlayOptions, Rendered, render_lesson, render_note
from ...teach.score import Lesson


class LessonTransport(QObject):
    positionChanged = pyqtSignal(float)      # lesson division; negative while counting in
    stateChanged = pyqtSignal(bool)          # playing?
    renderedChanged = pyqtSignal()           # a new render is ready (options changed)
    message = pyqtSignal(str)

    def __init__(self, parent=None) -> None:
        super().__init__(parent)
        self.backend = get_backend()
        self.cache = NoteCache()
        self.lesson: Lesson | None = None
        self.options = PlayOptions()
        self.loop = False
        self.rendered: Rendered | None = None
        self._dirty = True
        self._playing = False
        self._phase = "first"
        self._origin = 0.0
        self._offset = 0.0           # seconds into the buffer at which playback started
        self._position = 0.0         # last reported division
        self._timer = QTimer(self)
        self._timer.setInterval(33)
        self._timer.timeout.connect(self._tick)

    # ---------------------------------------------------------------- state

    @property
    def playing(self) -> bool:
        return self._playing

    @property
    def position(self) -> float:
        return self._position

    @property
    def available(self) -> bool:
        return bool(self.backend.available)

    def set_lesson(self, lesson: Lesson | None) -> None:
        self.stop()
        self.lesson = lesson
        self._position = 0.0
        self._dirty = True
        if lesson is not None:
            self.options = replace(self.options, tempo=float(lesson.tempo), loop_bars=None)
        self.renderedChanged.emit()
        self.positionChanged.emit(0.0)

    def update_options(self, **changes) -> None:
        """Change playback options; a running playback restarts from its current bar."""
        new = replace(self.options, **changes)
        if new == self.options:
            return
        self.options = new
        self._dirty = True
        self.renderedChanged.emit()
        if self._playing:
            self.play(from_division=self._bar_start(self._position), count_in=False)

    def set_loop(self, enabled: bool) -> None:
        self.loop = bool(enabled)

    # ------------------------------------------------------------ rendering

    def ensure_rendered(self) -> Rendered | None:
        if self.lesson is None:
            return None
        if self._dirty or self.rendered is None:
            self.rendered = render_lesson(self.lesson, self.options, self.cache)
            self._dirty = False
            for w in self.rendered.warnings:
                self.message.emit(w)
        return self.rendered

    def section_range(self) -> tuple[int, int]:
        if self.lesson is None:
            return 0, 0
        if self.options.loop_bars:
            return self.lesson.measure_range(*self.options.loop_bars)
        return 0, self.lesson.length

    def _bar_start(self, division: float) -> int:
        if self.lesson is None:
            return 0
        m = self.lesson.measure_divisions
        start, end = self.section_range()
        d = int(max(start, min(division, end - 1)) // m * m)
        return max(start, d)

    # ------------------------------------------------------------- playback

    def toggle(self) -> None:
        if self._playing:
            self.stop()
        else:
            self.play()

    def play(self, from_division: float | None = None, count_in: bool = True) -> None:
        rendered = self.ensure_rendered()
        if rendered is None:
            return
        if self._playing:
            self.backend.stop()
            self._timer.stop()
        start, _end = self.section_range()
        if from_division is None or from_division <= start:
            division = start
        else:
            division = self._bar_start(from_division)
        if count_in and division == start:
            offset = 0.0
            buffer = rendered.buffer
        else:
            offset = rendered.seconds_at(division)
            buffer = rendered.buffer
        self._phase = "first"
        self._offset = offset
        self._start_backend(buffer.slice_seconds(offset, buffer.duration) if offset > 0 else buffer, loop=False)
        self._origin = time.perf_counter()
        self._playing = True
        self._timer.start()
        self.stateChanged.emit(True)
        self._tick()

    def stop(self) -> None:
        if not self._playing:
            return
        self.backend.stop()
        self._timer.stop()
        self._playing = False
        start, _ = self.section_range()
        self._position = float(start)
        self.stateChanged.emit(False)
        self.positionChanged.emit(self._position)

    def seek(self, division: float) -> None:
        """Move the cursor; restarts playback from that bar if playing."""
        if self._playing:
            self.play(from_division=division, count_in=False)
        else:
            self._position = float(self._bar_start(division)) if self.lesson else 0.0
            self.positionChanged.emit(self._position)

    def preview_note(self, midi: int) -> None:
        if self.lesson is None or self._playing:
            return
        buf = render_note(self.lesson, midi, 0.7, self.options.sample_rate, self.cache)
        self._start_backend(buf.gain_db(self.options.voice_db + 4.0), loop=False)

    def _start_backend(self, buffer: AudioBuffer, loop: bool) -> None:
        if not self.backend.available:
            self.message.emit("No audio device — playback is silent, the cursor still moves")
            return
        try:
            self.backend.play(buffer, loop)
        except Exception as exc:
            self.message.emit(f"Playback failed: {exc}")

    # ---------------------------------------------------------------- clock

    def _tick(self) -> None:
        rendered = self.rendered
        if rendered is None or not self._playing:
            return
        elapsed = time.perf_counter() - self._origin
        if self._phase == "first":
            pos = self._offset + elapsed
            if pos >= rendered.buffer.duration:
                if self.loop:
                    self._phase = "loop"
                    self._origin = time.perf_counter()
                    self._start_backend(rendered.section, loop=True)
                    pos = rendered.count_in_seconds
                else:
                    self.stop()
                    return
        if self._phase == "loop":
            pos = rendered.count_in_seconds + (elapsed % max(1e-6, rendered.section.duration))
        self._position = rendered.division_at(pos)
        self.positionChanged.emit(self._position)

    def count_in_beat(self, division: float) -> int | None:
        """1-based beat number while the count-in is sounding, else None."""
        if self.lesson is None or self.rendered is None:
            return None
        start = self.rendered.start_division
        if division >= start:
            return None
        beats_before = (start - division) / self.lesson.beat_divisions
        per_bar = self.lesson.measure_divisions / self.lesson.beat_divisions
        total = self.options.count_in_bars * per_bar
        beat = int(total - beats_before) % int(per_bar) + 1
        return beat

File: web/src/transport.js
/* Playback with Web Audio. Position comes from AudioContext.currentTime, so
 * the cursor is sample-accurate. A looped practice range plays as two
 * stages, like the desktop app: count-in + first pass once, then the section
 * alone with loop=true, so the count-in is heard exactly once. */

import { PASSES, passAt, buildRunSamples } from "./run.js";

export class Transport {
  constructor(bridge) {
    this.bridge = bridge;
    this.ctx = null;
    this.index = null;
    this.lesson = null;
    this.options = { tempo: 80, hands: "both", metronome: true, count_in_bars: 1, loop_bars: null,
                     voice_db: -12, metronome_unpitched: false };
    this.onPass = null;          // called with the pass number each time a loop wraps
    this.pass = 0;
    this.run = null;             // a practice run in progress (see run.js)
    this.onRunPass = null;       // called with the pass index as the run moves on
    this.loop = false;
    this.rendered = null;          // {samples, sampleRate, frames, count_in_seconds, ...}
    this.buffer = null;            // AudioBuffer: count-in + section
    this.section = null;           // AudioBuffer: section only
    this.dirty = true;
    this.playing = false;
    this.phase = "first";
    this.source = null;
    this.startedAt = 0;
    this.offset = 0;
    this.position = 0;
    this.gain = null;
    this.onPosition = null;
    this.onState = null;
    this.onMessage = null;
    this.rendering = null;
    this._raf = null;
    this._generation = 0;
  }

  /* Must be called from a user gesture (click) the first time. */
  ensureContext() {
    if (!this.ctx) {
      this.ctx = new (window.AudioContext || window.webkitAudioContext)();
      this.gain = this.ctx.createGain();
      this.gain.connect(this.ctx.destination);
    }
    if (this.ctx.state === "suspended") this.ctx.resume();
    return this.ctx;
  }

  setLesson(index, lesson) {
    this.stop();
    this.index = index;
    this.lesson = lesson;
    this.options = { ...this.options, tempo: lesson ? lesson.tempo : 80, loop_bars: null };
    this.dirty = true;
    this.position = 0;
    this.emitPosition();
  }

  updateOptions(changes) {
    const next = { ...this.options, ...changes };
    if (JSON.stringify(next) === JSON.stringify(this.options)) return;
    this.options = next;
    this.dirty = true;
    if (this.playing) this.play(this.barStart(this.position), false);
  }

  setLoop(on) { this.loop = !!on; }

  sectionRange() {
    if (!this.lesson) return [0, 0];
    const m = this.lesson.measure_divisions;
    const lb = this.options.loop_bars;
    if (lb) {
      const first = Math.max(1, Math.min(lb[0], this.lesson.measures));
      const last = Math.max(first, Math.min(lb[1], this.lesson.measures));
      return [(first - 1) * m, last * m];
    }
    return [0, this.lesson.length];
  }

  barStart(division) {
    if (!this.lesson) return 0;
    const m = this.lesson.measure_divisions;
    const [start, end] = this.sectionRange();
    const d = Math.floor(Math.max(start, Math.min(division, end - 1)) / m) * m;
    return Math.max(start, d);
  }

  async ensureRendered() {
    if (this.index === null) return null;
    if (!this.dirty && this.rendered) return this.rendered;
    if (this.rendering) return this.rendering;
    const generation = ++this._generation;
    this.rendering = this.bridge.render(this.index, this.options).then((r) => {
      this.rendering = null;
      if (generation !== this._generation) return this.rendered;
      const ctx = this.ensureContext();
      const buffer = ctx.createBuffer(1, r.samples.length, r.sample_rate);
      buffer.copyToChannel(r.samples, 0);
      const countFrames = Math.round(r.count_in_seconds * r.sample_rate);
      const section = ctx.createBuffer(1, Math.max(1, r.samples.length - countFrames), r.sample_rate);
      section.copyToChannel(r.samples.subarray(countFrames), 0);
      this.rendered = r;
      this.buffer = buffer;
      this.section = section;
      this.dirty = false;
      for (const w of r.warnings || []) if (this.onMessage) this.onMessage(w);
      return r;
    });
    return this.rendering;
  }

  secondsAt(division) {
    const r = this.rendered;
    return r.count_in_seconds + (division - r.start_division) * r.seconds_per_division;
  }

  divisionAt(seconds) {
    const r = this.rendered;
    return r.start_division + (seconds - r.count_in_seconds) / r.seconds_per_division;
  }

  async play(fromDivision = null, countIn = true) {
    const r = await this.ensureRendered();
    if (!r) return;
    this.ensureContext();
    this._stopSource();
    const [start] = this.sectionRange();
    const division = fromDivision === null || fromDivision <= start ? start : this.barStart(fromDivision);
    const offset = countIn && division === start ? 0 : this.secondsAt(division);
    this.phase = "first";
    this.pass = 0;
    this.offset = offset;
    this._startSource(this.buffer, false, offset);
    this.playing = true;
    if (this.onState) this.onState(true);
    this._tick();
  }

  /* ------------------------------------------------------------- run */

  /* Render the three passes into one buffer. Returns the run, or null when
   * there is nothing to play. The caller may show progress while it works. */
  async prepareRun() {
    if (this.index === null || !this.lesson) return null;
    const ctx = this.ensureContext();
    const countIn = Math.max(1, this.options.count_in_bars || 0);
    const base = { ...this.options, count_in_bars: countIn, loop_bars: this.options.loop_bars };
    const full = await this.bridge.render(this.index, { ...base, voice_db: -12 });
    const clicks = await this.bridge.render(this.index, { ...base, voice_db: -100 });
    const rate = full.sample_rate;
    const countInFrames = Math.round(full.count_in_seconds * rate);
    // A pass is the count-in plus the section: a whole number of beats. What
    // the render has beyond that is the release tail, which belongs to the
    // next pass's count-in, not between the passes.
    const sectionSeconds = (full.end_division - full.start_division) * full.seconds_per_division;
    const passFrames = Math.round((full.count_in_seconds + sectionSeconds) * rate);
    const { samples } = buildRunSamples(full.samples, clicks.samples, countInFrames, passFrames);
    const buffer = ctx.createBuffer(1, samples.length, rate);
    buffer.copyToChannel(samples, 0);
    this.run = {
      buffer, rate, passFrames,
      passSeconds: passFrames / rate,
      countInSeconds: full.count_in_seconds,
      secondsPerDivision: full.seconds_per_division,
      startDivision: full.start_division,
      endDivision: full.end_division,
      index: -1,
      passes: PASSES,
    };
    this.rendered = full;          // the position mapping belongs to one pass
    return this.run;
  }

  /* Play a prepared run from its first pass. */
  playRun() {
    if (!this.run) return;
    this._stopSource();
    this.loopBeforeRun = this.loop;
    this.loop = false;
    this.phase = "run";
    this.run.index = -1;
    this.offset = 0;
    this._startSource(this.run.buffer, false, 0);
    this.playing = true;
    if (this.onState) this.onState(true);
    this._tick();
  }

  get inRun() { return !!this.run && this.phase === "run"; }

  endRun() {
    const had = !!this.run;
    this.run = null;
    if (this.loopBeforeRun !== undefined) { this.loop = this.loopBeforeRun; this.loopBeforeRun = undefined; }
    if (this.phase === "run") this.phase = "first";
    return had;
  }

  stop() {
    if (!this.playing) return;
    this._stopSource();
    this.playing = false;
    if (this._raf) cancelAnimationFrame(this._raf);
    this._raf = null;
    const [start] = this.sectionRange();
    this.position = start;
    if (this.onState) this.onState(false);
    this.emitPosition();
  }

  toggle() { if (this.playing) this.stop(); else this.play(); }

  seek(division) {
    if (this.playing) this.play(division, false);
    else {
      this.position = this.lesson ? this.barStart(division) : 0;
      this.emitPosition();
    }
  }

  async previewNote(midi) {
    if (this.index === null || this.playing) return;
    const ctx = this.ensureContext();
    const r = await this.bridge.preview(this.index, midi, 0.7);
    const buffer = ctx.createBuffer(1, r.samples.length, r.sampleRate);
    buffer.copyToChannel(r.samples, 0);
    const src = ctx.createBufferSource();
    src.buffer = buffer;
    src.connect(this.gain);
    src.start();
  }

  _startSource(buffer, loop, offset) {
    const src = this.ctx.createBufferSource();
    src.buffer = buffer;
    src.loop = loop;
    src.connect(this.gain);
    src.start(0, offset || 0);
    this.source = src;
    this.startedAt = this.ctx.currentTime;
  }

  _stopSource() {
    if (this.source) {
      try { this.source.stop(); } catch (_) { /* already stopped */ }
      this.source.disconnect();
      this.source = null;
    }
  }

  _tick() {
    if (!this.playing || !this.rendered) return;
    const elapsed = this.ctx.currentTime - this.startedAt;
    let pos;
    if (this.phase === "run") {
      const run = this.run;
      const t = this.offset + elapsed;
      if (t >= run.passSeconds * run.passes.length) { this.stop(); return; }
      const { index, offset } = passAt(t, run.passSeconds);
      if (index !== run.index) { run.index = index; if (this.onRunPass) this.onRunPass(index, run.passes); }
      this.position = this.divisionAt(offset);
      this.emitPosition();
      this._raf = requestAnimationFrame(() => this._tick());
      return;
    }
    if (this.phase === "first") {
      pos = this.offset + elapsed;
      if (pos >= this.buffer.duration) {
        if (this.loop) {
          this._stopSource();
          this.phase = "loop";
          this._startSource(this.section, true, 0);
          pos = this.rendered.count_in_seconds;
        } else {
          this.stop();
          return;
        }
      }
    }
    if (this.phase === "loop") {
      const e = this.ctx.currentTime - this.startedAt;
      const dur = Math.max(1e-6, this.section.duration);
      const pass = Math.floor(e / dur) + 1;
      if (pass !== this.pass) { this.pass = pass; if (this.onPass) this.onPass(pass); }
      pos = this.rendered.count_in_seconds + (e % dur);
    }
    this.position = this.divisionAt(pos);
    this.emitPosition();
    this._raf = requestAnimationFrame(() => this._tick());
  }

  emitPosition() { if (this.onPosition) this.onPosition(this.position); }

  countInBeat(division) {
    if (!this.lesson || !this.rendered) return null;
    const start = this.rendered.start_division;
    if (division >= start) return null;
    const beatsBefore = (start - division) / this.lesson.beat_divisions;
    const perBar = this.lesson.measure_divisions / this.lesson.beat_divisions;
    const total = this.options.count_in_bars * perBar;
    return (Math.floor(total - beatsBefore) % perBar) + 1;
  }
}
